java/src/main/java/com/powsybl/python/SensitivityAnalysisContext.java

```diff
@@ -64,84 +146,162 @@ void setBusVoltageFactorMatrix(List<String> busVoltageIds, List<String> targetVo
-                }
-            }
-
-            for (int row = 0; row < busVoltageMatrixRowCount; row++) {
-                final String targetVoltageId = targetVoltageEquipmentsIds.get(row);
-                for (int column = 0; column < busVoltageMatrixColCount; column++) {
-                    final String busVoltageId = busVoltageEquipmentsIds.get(column);
-                    handler.onFactor(SensitivityFunctionType.BUS_VOLTAGE, busVoltageId,
-                            SensitivityVariableType.BUS_TARGET_VOLTAGE, targetVoltageId, false, contingencyContext);
+                } else if (matrix.getFunctionType() == SensitivityFunctionType.BUS_VOLTAGE) {
+                    for (int row = 0; row < rows.size(); row++) {
+                        String targetVoltageId = rows.get(row);
+                        for (int column = 0; column < columns.size(); column++) {
+                            String busVoltageId = columns.get(column);
+                            handler.onFactor(SensitivityFunctionType.BUS_VOLTAGE, busVoltageId,
+                                    SensitivityVariableType.BUS_TARGET_VOLTAGE, targetVoltageId, false, ContingencyContext.all());
+                        }
+                    }
                 }
             }
         };
 
-        int baseCaseValueSize = branchFlowMatrixColumnCount * branchFlowMatrixRowCount + busVoltageMatrixColCount * busVoltageMatrixRowCount;
+        int baseCaseValueSize = getTotalNumberOfMatrixFactors(matrices);
         double[] baseCaseValues = new double[baseCaseValueSize];
         double[][] valuesByContingencyIndex = new double[contingencies.size()][baseCaseValueSize];
-        double[] baseCaseReferences = new double[branchFlowMatrixColumnCount + busVoltageMatrixColCount];
-        double[][] referencesByContingencyIndex = new double[contingencies.size()][branchFlowMatrixColumnCount + busVoltageMatrixColCount];
-        SensitivityValueWriter valueWriter = (factorIndex, contingencyIndex, value, functionReference) -> {
-            int columnIndex = factorIndex < busVoltageMatrixSerializedOffset ?
-                    factorIndex % branchFlowMatrixColumnCount
-                    : branchFlowMatrixColumnCount + (factorIndex - busVoltageMatrixSerializedOffset) % busVoltageMatrixColCount;
+
+        int totalColumnsCount = getTotalNumberOfMatrixFactorsColumns(matrices);
+        double[] baseCaseReferences = new double[totalColumnsCount];
+        double[][] referencesByContingencyIndex = new double[contingencies.size()][totalColumnsCount];
+
+        NavigableMap<Integer, MatrixInfo> factorIndexMatrixMap = new TreeMap<>();
+        for (MatrixInfo m : matrices) {
+            factorIndexMatrixMap.put(m.getOffsetData(), m);
+        }
+
+        SensitivityValueWriter valueWriter = (factorContext, contingencyIndex, value, functionReference) -> {
+            int factorIndex = factorContext;
+            MatrixInfo m = factorIndexMatrixMap.floorEntry(factorIndex).getValue();
```

Comment:
This is the most critical part for performance, because done for each of the huge number of factor. Maybe ok because small numer of matrices?

Reply:
I think usage should probably a few big matrices rather than lots of small matrices, in this cas this has minimal impact on performance. No way to enforce what usage will be though.